File: app/src/main/java/ru/zakoulov/gallery/imageController/tasks/LoadDailyImage.java
package ru.zakoulov.gallery.imageController.tasks;

import android.os.AsyncTask;
import android.util.Log;

import org.jsoup.Jsoup;
import org.jsoup.nodes.Document;
import org.jsoup.select.Elements;

import java.io.IOException;
import java.text.SimpleDateFormat;
import java.util.Calendar;
import java.util.GregorianCalendar;
import java.util.Locale;

import ru.zakoulov.gallery.imageController.Image;
import ru.zakoulov.gallery.imageController.ImageController;

/**
 * Created by Илья on 27.04.2018.
 */
public class LoadDailyImage extends AsyncTask<String, Void, Image> {
    Calendar calendar = null;
    TaskResponseDailyImage taskResponse;

    @Override
    protected Image doInBackground(String... params) {
        Image image = new Image();
        SimpleDateFormat sdf = new SimpleDateFormat("yyMMdd", Locale.getDefault());
        String path = params[0];
        boolean run = true;
        while (run) {
            try {
                run = false;
                if (calendar != null) // Выполняется в случае, когда идёт повторная итерация и надо загружать другую картинку
                    path = params[0] + "ap" + sdf.format(calendar.getTime()) + ".html";
                else
                    calendar = new GregorianCalendar();

                Document doc = Jsoup.connect(path).get();
                Elements elements = doc.body().select("center").first().select("p").get(1).select("a");

                String date = doc.body().select("center").first().select("p").get(1).text(); // Получение даты
                parseDate(date); // Парсинг даты

                if (elements.size() == 0) {
                    run = true;
                    calendar.add(Calendar.DAY_OF_YEAR, -1); // Переход к предыдущей картинке
                    continue;
                }
                image.setDate(calendar.getTime());
                image.setPath(elements.first().select("img").first().attr("src"));
                image.setFullPath(elements.first().attr("href"));
                image.setName(doc.body().select("center").get(1).select("b").first().text());
            } catch (IOException e) {
                Log.d("error", e.getMessage());
                taskResponse.responseDailyImageDownload(null);
                return null;
            }
        }
        taskResponse.responseDailyImageDownload(image);
        return image;
    }

    private void parseDate(String date) {
        String[] dates = date.split(" ");
        calendar.set(Calendar.YEAR, Integer.parseInt(dates[0]));
        int month = 0;
        // Не использую SimpleDateFormat т.к. не получается парсить месяц
        switch (dates[1]) {
            case "January":
                month = 0; break;
            case "February":
                month = 1; break;
            case "March":
                month = 2; break;
            case "April":
                month = 3; break;
            case "May":
                month = 4; break;
            case "June":
                month = 5; break;
            case "July":
                month = 6; break;
            case "August":
                month = 7; break;
            case "September":
                month = 8; break;
            case "October":
                month = 9; break;
            case "November":
                month = 10; break;
            case "December":
                month = 11; break;
        }
        calendar.set(Calendar.MONTH, month);
        calendar.set(Calendar.DAY_OF_MONTH, Integer.parseInt(dates[2]));
    }

    public void setTaskResponse(TaskResponseDailyImage taskResponse) {
        this.taskResponse = taskResponse;
    }
}

File: app/src/main/java/ru/zakoulov/gallery/imageController/tasks/LoadDailyImage.kt
package ru.zakoulov.gallery.imageController.tasks

import android.os.AsyncTask
import android.util.Log

import org.jsoup.Jsoup

import java.io.IOException
import java.text.SimpleDateFormat
import java.util.Calendar
import java.util.GregorianCalendar
import java.util.Locale

import ru.zakoulov.gallery.imageController.Image

class LoadDailyImage(val taskResponse: TaskResponseDailyImage) : AsyncTask<String, Void, Void>() {
    var calendar: Calendar = GregorianCalendar()

    override fun doInBackground(vararg params: String): Void? {
        val sdf = SimpleDateFormat("yyMMdd", Locale.getDefault())
        var path = params[0]
        var run = true
        var firstIteration = true
        lateinit var image: Image
        while (run) {
            try {
                run = false
                if (!firstIteration)
                    path = params[0] + "ap" + sdf.format(calendar.time) + ".html"
                firstIteration = false

                val doc = Jsoup.connect(path).get()
                val elements = doc.body().select("center").first().select("p")[1].select("a")

                val date = doc.body().select("center").first().select("p")[1].text() // Получение даты
                parseDate(date) // Парсинг даты

                if (elements.size == 0) {
                    run = true
                    calendar.add(Calendar.DAY_OF_YEAR, -1) // Переход к предыдущей картинке
                    continue
                }

                image = Image(
                        date=calendar.time,
                        path=elements.first().select("img").first().attr("src"),
                        fullPath=elements.first().attr("href"),
                        name=doc.body().select("center")[1].select("b").first().text()
                )
            } catch (e: IOException) {
                Log.d("error", e.message)
                taskResponse.responseDailyImageError()
                return null
            }

        }
        taskResponse.responseDailyImageDownload(image)
        return null
    }

    /** Из строки date сохраняет дату в calendar
     * @param date Строка с сайта, в которой написана дата публикации картинки
     */
    private fun parseDate(date: String) {
        val dates = date.split(" ")
        calendar.set(Calendar.YEAR, Integer.parseInt(dates[0]))
        var month = 0
        // Не использую SimpleDateFormat т.к. не получается парсить месяц
        when (dates[1]) {
            "January" -> month = 0
            "February" -> month = 1
            "March" -> month = 2
            "April" -> month = 3
            "May" -> month = 4
            "June" -> month = 5
            "July" -> month = 6
            "August" -> month = 7
            "September" -> month = 8
            "October" -> month = 9
            "November" -> month = 10
            "December" -> month = 11
        }
        calendar.set(Calendar.MONTH, month)
        calendar.set(Calendar.DAY_OF_MONTH, Integer.parseInt(dates[2]))
    }
}
